### DISCUSSION TITLE
Create new resources

### DISCUSSION THREAD

**Discussion by pepijndik** [lucasgiovanny/laravel-prestashop#5] [Ideas]
Currently, it's only possible to GET all the resources from Prestashop.
It would be very nice to have also the options to save() and create new resources.

I would start a PR for this and hopes it turns out nice :D

**Comment by lucasgiovanny** [comment]
That's awesome! It was my idea since the beginning. Let's keep in touch, I can definitely help with that.

**Reply by pepijndik** [reply]
Will keep you updated,
I had started first also my own repo but i thought ist better to combine forces 😁 with the love of open source
I have some model function inside of it maybe idea to use these? What do you think?

https://github.com/pdik/LaravelPrestashop

**Reply by lucasgiovanny** [reply]
I'll invite you to collaborate on the repo so that we can start coding a new branch for those features. What do you think? You can bring what you have done before so we join forces here.

**Reply by pepijndik** [reply]
Sounds nice!

**Reply by pepijndik** [reply]
This weekend I have worked on it, And have joined both ends together.
Will make a PR request to the v2-dev branch

**Reply by pepijndik** [reply]
Can you add me to the docu repo, So I can make some changes there?
Have just as we speak made a PR and merged it to the v2-dev branch.
There are some nice things added.  you can retrieve the resources from your way with the Prestashop facade like `prestashop::orders()`
But its now also possible to first create one connection and work directly from the resource end like so:
`            $con = new PrestashopClient();
            $order = new Orders($con)->get();`  
I have also grouped everything that belongs into a findable trait like the where, sort, sortby, etc
Also, i made a basic start for the Put and post request. i thought that the Prestashop web service only accepts XML for POST and PUT so therefore have to make a parser function,  
`createXmlFromModel()` //Will create a XML from the resource that's given. 
`parseArrayToXml()`  //Parses a basic array to XML used inside the xmlFromModel

**Reply by pepijndik** [reply]
We have to add a lot more 🧾resources but I have added some basics🔥

**Reply by pepijndik** [reply]
Working btw on some tests 😄

**Reply by lucasgiovanny** [reply]
That's awesome! I just added you to the docs repo. I'll create there today a Github Action for automatic deployment so we can have it sync with the live website. Please add yourself as an author on the doc as well :)

I'll have a look later today at your branch, I believe I can contribute with more resources today.

**Reply by lucasgiovanny** [reply]
About your PR (#4), I believe that we can work together on the branch that you created and then release everything together as version 1.0. I'll wait to merge everything to the main.

**Reply by pepijndik** [reply]
sounds nice, i can do more later today :D  Will try to get the POST request working. but i am struggling in my head with the relations from the models

**Reply by pepijndik** [reply]
is a better idea and much more version control able,
When I made that PR I didn't know yet that i was going to work more on this

**Reply by pepijndik** [reply]
And it seems like that my first thoughts where right, the  Presta webservice only allows XML post and put request.

https://github.com/PrestaShop/PrestaShop/blob/develop/webservice/dispatcher.php

**Reply by pepijndik** [reply]
Maybe an idea to add badges for the future I don't know yet how to set up these but i think they work on test

**Reply by pepijndik** [reply]
What is the main idea of the were function? right now in the dev branch it's not possible to add multiple wheres,
Should we extend it so it will  support this?

**Reply by lucasgiovanny** [reply]
Once we finish the branch for the new release, I will add Github badges for testing and some automation to have this covered.

**About the where method**
The idea is to be able to add as many `wheres` as you want, just like Laravel Eloquent. The method itself it's a shortcut for the `filter` method. The idea of the package since the beginning is to have a close syntax to Eloquent, so that's why I add this method. We can keep it.

```php
// Example of using it
Prestashop::orders()->where('date_add', 'INTERVAL', ['2022-01-01', '2022-02-01'])
    ->where('order_state_id', 2)
    ->get();

// The "where" method
public function where(string $field, string $operatorOrValue, $value = null)
    {
        return $this->filter($field, $operatorOrValue, $value);
    }

// Shortcut to filter method
public function filter(string $field, string $operatorOrValue, $value = null)
    {
        $operator = $value ? $operatorOrValue : '=';

        if (! in_array(strtoupper($operator), self::FILTER_OPERATORS)) {
            throw new Exception('Invalid filter operator');
        }

        $this->filters[] = [
            'field' => $field,
            'operator' => strtoupper($operator),
            'value' => $value ?: $operatorOrValue,
        ];

        return $this;
    }
```

**Reply by pepijndik** [reply]
Oke nice then i  have understand it the right way **😅**

**Comment by pepijndik** [comment]
Have just fixed same small parts to make the first Post request for the Orders resource,
have also implemented a validation to validate the required parts because the web service won't give a errors for this

To test and use:
```php
       $order = PrestashopClient::orders();
        $order->id_address_delivery = 1;
        $order->id_address_invoice = 1;
        $order->id_cart = 1;
        $order->id_currency = 1;
        $order->id_lang = 1;
        $order->id_customer = 4;
        $order->id_carrier = 1;
        $order->module = "Cheque";
        $order->payment = "ps_checkpayment";
        $order->total_paid = 0;
        $order->total_paid_real = 0;
        $order->total_products = 1;
        $order->total_products_wt = 1;
        $order->conversion_rate = 1;
        $order->save();`

**Reply by pepijndik** [reply]
Just notice that the server is responding with a 500 error,
And not sure why because the error message is empty... 😢 
It seems that is because of a strange Prestashop thing ... When i change in the defines.inc on the Prestashop side:
and set define('_PS_MODE_DEV_', false); to true,  then i am able again to get a  blank XML back 🎆

**Comment by pepijndik** [comment]
🥳 @lucasgiovanny  Finally I have made it possible to create, insert and save new resources.
Right now I have tested this for 2 resources, the carts, and addresses. But I am sure if we have added all resources were able to create all resources. 
Because the Carts and orders are the hardest 💯 ones.

**Reply by pepijndik** [reply]
Have also fixed some validation on the models to prevent some nasty XML webservice errors.
i have tried to parse those errors to make them readable for now for me the way i have it now it works but it can be better i think :D

**Comment by pepijndik** [comment]
I think, where almost ready to make from here a merge request for v2, The only thing i can think of where missing is images. But this is not that big issue i think
